blockchain/upgrade.go

```diff
@@ -3363,6 +3418,293 @@ func separateUtxoDatabase(ctx context.Context, db database.DB, utxoDb database.D
 	return db.Flush()
 }
 
+// fetchLegacyBucketId returns the legacy bucket id for the provided bucket
+// name.  A Get function must be provided to retrieve key/value pairs from the
+// underlying data store.
+func fetchLegacyBucketId(getFn func(key []byte) ([]byte, error),
+	bucketName []byte) ([]byte, error) {
+
+	// bucketIndexPrefix is the prefix used for all entries in the bucket index.
+	bucketIndexPrefix := []byte("bidx")
+
+	// parentBucketId is the parent bucket id.  This is hardcoded to zero here
+	// since none of the legacy UTXO database buckets had parent buckets.
+	parentBucketId := []byte{0x00, 0x00, 0x00, 0x00}
+
+	// Construct the key and fetch the corresponding bucket id from the database.
+	// The serialized bucket index key format is:
+	//   <bucketindexprefix><parentbucketid><bucketname>
+	bucketIdKey := prefixedKey(bucketIndexPrefix, parentBucketId)
+	bucketIdKey = append(bucketIdKey, bucketName...)
+	bucketId, err := getFn(bucketIdKey)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching legacy bucket id for %v bucket: %v",
+			string(bucketName), err)
+	}
+
+	return bucketId, nil
+}
+
+// migrateUtxoDbBuckets migrates the UTXO data to use simple key prefixes rather
+// than buckets.
+func migrateUtxoDbBuckets(ctx context.Context, utxoBackend UtxoBackend) error {
+	// Key names, versions, and prefixes are hardcoded below so that updates do
+	// not affect old upgrades.
+
+	// UTXO key sets and versions.
+	const utxoKeySetDbInfoV1 = 1
+	const utxoKeySetUtxoStateV1 = 2
+	const utxoKeySetUtxoSetV3 = 3
+	const utxoKeySetDbInfoVersion = 1
+	const utxoKeySetUtxoStateVersion = 1
+	const utxoKeySetUtxoSetVersion = 3
+
+	// Legacy buckets.
+	metadataLegacyBucketID := []byte{0x00, 0x00, 0x00, 0x00}
+	utxoSetLegacyBucketName := []byte("utxosetv3")
+	utxoDbInfoLegacyBucketName := []byte("dbinfo")
+	legacyBucketIndexPrefix := []byte("bidx")
+
+	// Database info keys.
+	utxoPrefixDbInfoV1 := []byte{utxoKeySetDbInfoV1, utxoKeySetDbInfoVersion}
```

Comment:
You are correct that the old bucketized version information should not be removed, and instead, the version should be bumped.  I'll get that updated and test the downgrade paths, thanks for pointing that out.

For the new scheme, what do you think about leaving the version for the database info key set as-is, but adding documentation that if the version _is_ ever bumped for that key set, then the old version should also remain in the database, indicating that a newer version exists (so that the old code can throw an error).  I'm good with your suggestion as well, but I kind of like keeping it consistent with the structure of the other key sets, and it would allow for a safe way of updating that ket set if needed in the future.

Reply:
That seems fine.  I personally find it a bit confusing to have the version information versioned, but I understand the desire to keep it consistent and that seems reasonable.  As long as it's documented exactly what the upgrade process must look like in order to ensure old versions of the software continue to properly detect and reject newer database versions, I don't see any issue with that.

Reply:
I may go with something like this, which is closer to what you suggested but just using `0` rather than `m`:

```go
// utxoKeySetNoVersion defines the value to be used for the version of key sets
// where versioning does not apply.
const utxoKeySetNoVersion = 0

// utxoKeySetVersions defines the current version for each UTXO backend key
// set.
var utxoKeySetVersions = map[utxoKeySet]uint8{
	// Note: The database info key set must remain at fixed keys so that older
	// software can properly load the database versioning info, detect newer
	// versions, and throw an error.
	utxoKeySetDbInfo:    utxoKeySetNoVersion,
	utxoKeySetUtxoState: 1,
	utxoKeySetUtxoSet:   3,
}
```

Reply:
That looks pretty good and seems clear.  The upgrade code can then just do `utxoDbInfoVersionKeyNew := []byte("\x01\x00version")` which should make that part a little easier to follow too.

As a bit of an aside, since this code comes before the code over in `utxobackend.go` that describes the way it works in the diff, when I was reviewing it, I was initially confused by the `utxoPrefixDbInfoV1` and `utxoSetStateKeyNameV1`  since my initial impression was that it was migrating **to** version 1 of those things given it builds a target key, but I knew it was actually migrating to version 2 of the database.  Once I skipped forward to go look at the new format it made sense, but that is why I made the suggestion of changing the variable names in there to `utxoDbInfoVersionKeyX := /* hard coded target key name */` which also avoids a call to `prefixedKey` which is "production" code outside of the upgrade path and could theoretically change out from under the upgrade.  This is obviously not a big deal, but just thought I'd point out my thought process behind the suggestion there.

Reply:
That makes sense.  The main reason I defined the variables such as `utxoDbInfoVersionKeyNameV1` separately, rather than having hardcoded target key names as suggested, is that those variables are needed for the old keys as well (and since the old keys are prefixed by a dynamically defined bucket id, they can't really be hardcoded).  I could see how all of those key name definitions in the upgrade code might be hard to follow at first, though, so I'll see if I can make that a bit more clear.

Reply:
The version for the database info was updated to use `const utxoKeySetNoVersion = 0` as noted in my prior comment.  I also reworked how the key names are defined in the upgrade code to make it more self-explanatory with a descriptive comment and hard-coded keys, as you suggested.

Reply:
Regarding downgrades, they are working as expected (i.e., you get an error that the newer database version is not compatible), with one caveat.  Downgrading from this to one of the commits during the transitionary phase of separating the UTXO database won't display a proper error message due to a couple of reasons:

- If the UTXO database was already moved by running the code in this PR, the old code will (incorrectly) recreate it in the old location resulting in an incomplete UTXO set
- If the old code _does_ try to open the UTXO database with the `database` package, it will fail due to the write cursor not existing in the database

Both of these are solvable, but I first wanted to get an opinion on whether it is even worth adding additional code for it.  The reason being this will only occur if you are trying to downgrade between a handful of commits on `master` that were never released.

Reply:
I think it's fine to punt on a handful of commits that were never released here.  The main concern there is the case where you're trying to use `git bisect` to find when some issue was introduced and when checking out PRs to test.

For the first case, similar types of issues happen with the normal database there where the older version will intentionally fail to load.  That's why it's much more common to use `simnet` which nukes the db each time when doing that type of testing.

For the second case, as long as we make sure there are no PRs that fall in that range, and rebase them if they do, I don't think it will be an issue.